Amazon.QLDB.Driver.Tests/QldbDriverTests.cs

```diff
@@ -222,15 +227,92 @@ public void TestExecuteWithFuncLambdaReturnsFuncOutput()
         }
-        
+
+        [DataTestMethod]
+        [DynamicData(nameof(CreateDriverExceptions), DynamicDataSourceType.Method)]
+        public void Execute_ExceptionOnExecute_ShouldOnlyRetryOnISEAndTAOE(Exception exception, bool expectThrow)
+        {
+            var statement = "DELETE FROM table;";
+            var h1 = QldbHash.ToQldbHash("transactionId");
+            h1 = Transaction.Dot(h1, statement, new List<IIonValue> { });
+
+            var sendCommandResponseWithStartSession = new SendCommandResponse
+            {
+                StartTransaction = new StartTransactionResult
+                {
+                    TransactionId = "transactionId"
+                },
+                ResponseMetadata = new ResponseMetadata
+                {
+                    RequestId = "testId"
+                }
+            };
+
+            var sendCommandResponseExecute = new SendCommandResponse
+            {
+                ExecuteStatement = new ExecuteStatementResult
+                {
+                    FirstPage = new Page
+                    {
+                        NextPageToken = null
+                    }
+                },
+                ResponseMetadata = new ResponseMetadata
+                {
+                    RequestId = "testId"
+                }
+            };
+
+            var sendCommandResponseCommit = new SendCommandResponse
+            {
+                CommitTransaction = new CommitTransactionResult
+                {
+                    TransactionId = "transactionId",
+                    CommitDigest = new MemoryStream(h1.Hash),
+                },
+                ResponseMetadata = new ResponseMetadata
+                {
+                    RequestId = "testId"
+                }
+            };
+            var mockCreator = new Mock<Func<Session>>();
+            var mockSession = new Mock<Session>(null, null, null, null, null);
+
+            mockSession.Setup(x => x.StartTransaction()).Returns(sendCommandResponseWithStartSession.StartTransaction);
+            mockSession.SetupSequence(x => x.ExecuteStatement(It.IsAny<String>(), It.IsAny<String>(), It.IsAny<List<IIonValue>>()))
+                .Throws(exception)
+                .Throws(exception)
+                .Returns(sendCommandResponseExecute.ExecuteStatement);
+            mockSession.Setup(x => x.CommitTransaction(It.IsAny<String>(), It.IsAny<MemoryStream>()))
+                .Returns(sendCommandResponseCommit.CommitTransaction);
+
+            mockCreator.Setup(x => x()).Returns(mockSession.Object);
+
+            var driver = new QldbDriver(
+                new SessionPool(mockCreator.Object, 4, 4, NullLogger.Instance), NullLogger.Instance);
+            try
+            {
+                driver.Execute(txn => txn.Execute(statement));
+            }
+            catch (Exception e)
+            {
+                Assert.IsTrue(expectThrow);
+                Assert.AreEqual(exception, e);
```

Comment:
if it's retrying on the exceptions then why are they getting bubbled up?

Reply:
It would only throw when 'expectThrow' is true. Otherwise the test will fail if it throws.